### DISCUSSION TITLE
Batch component updates

### DISCUSSION THREAD

**Discussion by stancl** [livewire/livewire#3711] [3. Feature ideas]
Currently Livewire does not support updating multiple components using a single HTTP request. This would be a very significant addition for performance, considering that **90%+ of Livewire's "slowness" (compared to SPA solutions) comes from the HTTP overhead and framework boot-up**, rather than the actual Livewire code either on the backend or the frontend.

Aside from performance in edge cases, this would add a huge feature to Livewire: **the ability to split pages into multiple components**. Vue.js style.

Right now we need to put everything into a single component. Sure we can use Blade components and traits for the PHP part of it, but we can't really compose Livewire components of child Livewire components that **encapsulate both presentation and logic - because of the performance downsides**.

My suggestion is to add a new route that accepts an array of component updates and returns an array of component responses.

Then there would be a small change on the frontend to support this data structure (it's really the same as before, but now wrapped in `[]`).

This means that if you got a response from the server calling e.g. `billingUpdated` and you had 5 components listening to that event, Livewire wouldn't fire 5 separate requests which have their own HTTP overhead, but instead a single request with:

```
[
    { "serverMemo": ..., "changes": {...} },
    { "serverMemo": ..., "changes": {...} },
    { "serverMemo": ..., "changes": {...} },
]
```

And receive:

```
[
    { "effects": ..., "serverMemo": {...} },
    { "effects": ..., "serverMemo": {...} },
    { "effects": ..., "serverMemo": {...} },
]
```

## Examples

- A billing page where each part of it is a Livewire component, and e.g. updating the default payment method should update the subscription component, then the subscription component's changes should be reflected in the message at the top of the page https://twitter.com/samuelstancl/status/1277920792991449088
- in Lean Admin I use `$refresh` to update components when the data displayed in them (e.g. fields for some models) is modified in another component. `$refresh` is used instead of listeners because I have a custom system for detecting these changes

## Implementation

**Backend:**
1. Create a new route to handle bulk component updates
2. Make it accept essentially what `/livewire/foo` accepts, but as an array. So it'd send an array of payloads that are normally sent to `/livewire/foo` and `/livewire/bar`.
3. Make the route loop through the payloads and return an array of Livewire responses
4. (Improvement) check all responses for fired events, and if they contain an event that another component in this request listens to, run the listener method in the component

**Frontend:**
1. Replace the direct [fetch in `sendMessage()`](https://github.com/livewire/livewire/blob/69575f50bb7f8a49a41f9bd6bd16c73a6ef4fda3/js/connection/index.js#L32) with a call to a new abstraction which will push the message to an array
2. Each `sendMessage()` should also schedule a call that will send the actual request at the end of the event loop. I don't understand the event loop perfectly, but I think that this should batch the requests correctly in most cases. Also the benefit is that if the batching didn't happen for some reason, the normal `send()` was still scheduled so the component will work just fine and with zero practical delay.

```js
// in sendMessage()
messageScheduler.pushMessage(message)

setTimeout(() => messageScheduler.send(), 1);

// in MessageScheduler
send () {
    if (! this.messages.length) return; // send() will get scheduled multiple times, but should only make requests when it has data to work with

    if (this.messages.length === 1) this.sendOne(this.messages[0]) // this is a normal LW request, should use the /livewire/foo route

    // otherwise:
    // fetch() to the batch update route
}

**Comment by aneeskhan47** [comment]
@calebporzio

**Comment by phh** [comment]
Guess this solution could also be a way of adding a store of "global state", like with stores.

Following! Really interested in this one.

**Comment by aspyropoulos** [comment]
I have also opened a [discussion](https://github.com/livewire/livewire/discussions/3672) about that and I am glad to see that others also are interested in solving it. It would be nice to have a solution like this in the upcoming releases and get rid of the extra HTTP overhead.

**Comment by calebporzio** [comment]
Thanks for the writeup @stancl. I few thoughts:

I agree this would be a good optimization.

Because the requests currently would be happening in parallel, the extra framework booting and HTTP overhead would be shared right?

I'd love to see a benchmark comparison of the speeds of these two approaches. I'm almost tempted to say batching the requests into one request would be slower?

But again, before we sink time and resources into this, I'd love to see a benchmark on it so we know the gains. Thanks

**Reply by stancl** [reply]
> Because the requests currently would be happening in parallel, the extra framework booting and HTTP overhead would be shared right?

That depends. Say that this this happens:
- `PaymentMethod` component changed -> response fires `updateBilling`
- `updateBilling` triggers an update of `BillingBanner`, `BillingAddress`, `BillingPlan`, `BillingEmail`
- 4 separate requests are sent. Each one takes eg. 100ms network overhead, 50ms framework bootup, 10ms execution
- 4 separate responses are received at slightly different times. If we block the page, it'll unblock upon the reception of the last request. If we don't block it it'll be a bit jumpy with things getting updated at different times.

If we just fired one request, we could show a loading spinner and unblock the billing section of the page when the response is received.

So even though this is on first sight a performance improvement, it's also about allowing proper componentization of pages. I think it kind of allows that mental model of using Livewire. Due to the quirkiness of separate requests & responses doing UI updates, I'd tend to avoid making separate LW components and would just go with one massive component.

Also, in some cases processing the updates in a single request could prevent requests that'd happen *afterwards*. For example, say that we're updating `BillingPlan` and `BillingAddress`. One of those components fire an event that executes a method in the other component. This would normally be done in a separate requests. So a single change can actually call the same components multiple times. But here we'd notice that we have both components available, so we can run updates on the components multiple times if more events get fired during execution.

But again, I think that **the main value is in receiving a single response, without having to block the UI until the last request is received**. If there's a 20% chance that a request will get random extra latency, it'll happen much more likely when we're firing 5 requests than when we're firing one. The risk of random slowdowns becomes minimal.

So for instance the example in Lean — I DM'd you how the entire UI updates when some model's value changes. That's all separate requests. If it simply worked like this:
- The user makes a change
- The change is saved
- Then, a few moments later, the rest of the UI updates — at once — to reflect the new change

then that'd feel a lot smoother I think.

So it's more about getting that **single page feel for the users, while letting developers use separate components**. That improves both UX and DX greatly. And the benefit looking strictly at performance is avoiding random request slowdowns.

I agree with your thought that theoretically running the requests in parallel would be faster, which is why it should remain an option to manage updates using parallel requests (it makes more sense for some UIs). But in general, the overhead of processing the LW component will be lower than the network overhead, which makes the "single/united page feel" worth it I think.

**Reply by ragulka** [reply]
I agree with @stancl fully here, and for me personally, this is what stops me from using Livewire for anything non-trivial. The chain of requests that may fire across different components listening to events will eventually lead to "jumpy" UI. This is especially true if you try to update the UI as the user is typing, for example.

I like the concept of [Turbo Frames](https://turbo.hotwired.dev/handbook/frames). It supports multiple component updates within a single response.

Edit: stuffing everything into a single Livewire component is not really a feasible route.

**Comment by ariaieboy** [comment]
The good news is in livewire 3 this gonna be the default behavior in the liveiwre
